public/app/features/dashboard/components/DashboardPrompt/DashboardPrompt.tsx

```diff
@@ -165,7 +165,7 @@ export function ignoreChanges(current: DashboardModel, original: object | null)
 /**
  * Remove stuff that should not count in diff
  */
-function cleanDashboardFromIgnoredChanges(dashData: unknown) {
+function cleanDashboardFromIgnoredChanges(dashData: Dashboard) {
```

Comment:
Heyo 👾 After doing a mob session on fixing this error (`dashData` cannot be unknown, as it is passed to `DashboardModel` constructor which expects a `Dashboard`), we realised that there is an issue with [the getSaveModelClone() of DashboardModel](https://github.com/grafana/grafana/blob/use-veneer-dashboard-model/public/app/features/dashboard/state/DashboardModel.ts#L223). It returns a DashboardModel _type_, but in reality it does not return a new instance of the `class DashboardModel`, but a plain object with the properties in a Dashboard. Changing the return value to `Dashboard` creates much more errors everywhere else. We are wondering which way is best to go from here.

The `DashboardModel` has a lot more required fields than the `Dashboard` interface created by Cue does. We are wondering what the constraints are here about the json of a dashboard in the frontend vs. in the backend? @torkelo maybe you can answer this or point us in the right direction? Is the frontend supposed to be much more constrained than the backend when creating a new dashboard, or can the Cue schema of the dashboard have much more required fields set in itself?

Reply:
i tried changing the return type of `getSaveModelClone` to the generated `Dashboard` type, and looked over a bunch of the errors. It seems like there are probably a few techniques that need to be applied.

But, before we get into those - would it be reasonable to defer those fixes to a follow-up PR? This is such a large task, it seems to me that the more we can break it down into stages, the better.

Reply:
> The `DashboardModel` has a lot more required fields than the `Dashboard` interface created by Cue does.

We've [discussed this a bit in slack](https://raintank-corp.slack.com/archives/C03BH67DS2J/p1668528266330109) already, and its part of a larger, tricky topic - defaults and field optionality.

The key thing we need to start getting at is what it means for there to be a default value, vs. an optional field. I talked about that a bit in [this CUE issue](https://github.com/cue-lang/cue/discussions/1951#discussioncomment-3969260). 

But, concretely -

> or can the Cue schema of the dashboard have much more required fields set in itself?

In general, my guess is that it should be fine to make more fields required in the schema, WITH a default value.

> Is the frontend supposed to be much more constrained than the backend when creating a new dashboard

The main audience here to consider is probably more the as-code use cases, rather than the backend. Outside of public dashboards, there's little 'backend' code that makes a dashboard. When we think about that audience in the context of optional/default fields, it's pretty similar to what we think about in the frontend - what's the minimum the user _must_ state in an object in order to be a basically valid dashboard?

Reply:
@polibb yea, the getSaveModelClone returns the persisted state DTO so not DashboardModel

>  Is the frontend supposed to be much more constrained than the backend when creating a new dashboard

No, DashboardModel constructor is super forgiving, nothing is required as it applies defaults